contracts/solidity/NFTXYieldStakingZap.sol

```diff
@@ -0,0 +1,336 @@
+    address lpToken = pairFor(baseToken, address(WETH));
+    IERC20Upgradeable(lpToken).safeApprove(address(lpStaking), liquidity);
+    lpStaking.timelockDepositFor(vaultId, msg.sender, liquidity, 48 hours);
+    
+    // Return any token dust to the caller
+    uint256 remainingTokens = vaultTokenAmount - amountToken;
+    if (remainingTokens != 0) {
+      IERC20Upgradeable(baseToken).transfer(msg.sender, remainingTokens);
+    }
+
+    // Return any left of WETH to the user as ETH
+    uint256 remainingWETH = WETH.balanceOf(address(this)) - wethBalance;
+    if (remainingWETH > 0) {
+      // Unwrap our WETH into ETH and transfer it to the recipient
+      WETH.withdraw(remainingWETH);
+      (bool success, ) = payable(msg.sender).call{value: remainingWETH}("");
+      require(success, "Unable to send unwrapped WETH");
+    }
+  }
+
+
+  /**
+   * @notice Calculates the CREATE2 address for a sushi pair without making any
+   * external calls.
+   * 
+   * @return pair Address of our token pair
+   */
+
+  function pairFor(address tokenA, address tokenB) internal view returns (address pair) {
+    (address token0, address token1) = sortTokens(tokenA, tokenB);
+    pair = address(uint160(uint256(keccak256(abi.encodePacked(
+      hex'ff',
+      sushiRouter.factory(),
+      keccak256(abi.encodePacked(token0, token1)),
+      hex'e18a34eb0e04b04f7a0ac29a6e80748dca96319b42c54d679cb821dca90c6303' // init code hash
+    )))));
+  }
+
+
+  /**
+   * @notice Returns sorted token addresses, used to handle return values from pairs sorted in
+   * this order.
+   */
+
+  function sortTokens(address tokenA, address tokenB) internal pure returns (address token0, address token1) {
+      require(tokenA != tokenB, 'UniswapV2Library: IDENTICAL_ADDRESSES');
+      (token0, token1) = tokenA < tokenB ? (tokenA, tokenB) : (tokenB, tokenA);
+      require(token0 != address(0), 'UniswapV2Library: ZERO_ADDRESS');
+  }
+
+
+  /**
+   * @notice Allows our owner to withdraw and tokens in the contract.
+   * 
+   * @param token The address of the token to be rescued
+   */
+
+  function rescue(address token) external onlyOwner {
+    if (token == address(0)) {
+      (bool success, ) = payable(msg.sender).call{value: address(this).balance}("");
+      require(success, "Address: unable to send value");
+    } else {
+      IERC20Upgradeable(token).safeTransfer(msg.sender, IERC20Upgradeable(token).balanceOf(address(this)));
+    }
+  }
+
+
+  /**
+   * @notice Swaps ERC20->ERC20 tokens held by this contract using a 0x-API quote.
+   *
+   * @param buyToken The `buyTokenAddress` field from the API response
+   * @param swapCallData The `data` field from the API response
+   */
+
+  function _fillQuote(
+    address buyToken,
+    bytes calldata swapCallData
+  ) internal returns (uint256) {
+      // Track our balance of the buyToken to determine how much we've bought.
+      uint256 boughtAmount = IERC20Upgradeable(buyToken).balanceOf(address(this));
```

Comment:
Why is this handled using contract state instead of just operating based on the inputs?